src/khoj/processor/conversation/gpt4all/chat_model.py

```diff
@@ -0,0 +1,137 @@
+from typing import Union, List
+from datetime import datetime
+import sys
+import logging
+from threading import Thread
+
+from langchain.schema import ChatMessage
+
+from gpt4all import GPT4All
+
+
+from khoj.processor.conversation.utils import ThreadedGenerator, generate_chatml_messages_with_context
+from khoj.processor.conversation import prompts
+from khoj.utils.constants import empty_escape_sequences
+
+logger = logging.getLogger(__name__)
+
+
+def extract_questions_falcon(
+    text: str,
+    model: str = "ggml-model-gpt4all-falcon-q4_0.bin",
+    loaded_model: Union[GPT4All, None] = None,
+    conversation_log={},
+    use_history: bool = False,
+    run_extraction: bool = False,
+):
+    """
+    Infer search queries to retrieve relevant notes to answer user query
+    """
+    all_questions = text.split("? ")
+    all_questions = [q + "?" for q in all_questions[:-1]] + [all_questions[-1]]
+    if not run_extraction:
+        return all_questions
+
+    gpt4all_model = loaded_model or GPT4All(model)
+
+    # Extract Past User Message and Inferred Questions from Conversation Log
+    chat_history = ""
+
+    if use_history:
+        chat_history = "".join(
+            [
+                f'Q: {chat["intent"]["query"]}\n\n{chat["intent"].get("inferred-queries") or list([chat["intent"]["query"]])}\n\nA: {chat["message"]}\n\n'
+                for chat in conversation_log.get("chat", [])[-4:]
+                if chat["by"] == "khoj"
+            ]
+        )
+
+    prompt = prompts.extract_questions_falcon.format(
+        chat_history=chat_history,
+        text=text,
+    )
+    message = prompts.general_conversation_falcon.format(query=prompt)
+    response = gpt4all_model.generate(message, max_tokens=200, top_k=2)
```

Comment:
Wouldn't a `top_k=2` slow the generation down? Is this done to improve the quality of the question extraction response?

Reply:
See documentation for `top_k` parameter [here](https://docs.gpt4all.io/gpt4all_python.html#chatting-with-gpt4all).

It samples from next `k` predicted tokens in the response. I do think it helps with output quality, though I can't say for certain.